File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Backtracking/46_Permutations.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-19 15:29:06
 */

c.. Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        dfs(nums, 0, ans);
        r_ ans;
    }

    void dfs(vector<int> &num, int begin, vector<vector<int>> &ans){
        __(begin __ num.size()){
            ans.push_back(num);
            r_;
        }
        ___(int i=begin; i<num.size();i++){
            swap(num[begin], num[i]);
            dfs(num, begin+1, ans);
            // Backtrack
            swap(num[begin], num[i]);
        }
    }
};

/*
[]
[1]
[1,2,3]
*/

File: 125_algorithms/_exercises/templates/_algorithms_challenges/leetcode/leetCode/Backtracking/46_Permutations.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


c.. Solution o..
    # Easy to understand: recursively.
    ___ permute  nums
        ans   # list
        self.dfs(nums, [], ans)
        r_ ans

    ___ dfs  nums, path, ans
        __ n.. nums:
            ans.a.. path)
        ___ i, n __ enumerate(nums
            self.dfs(nums[:i] + nums[i + 1:], path + [n], ans)


c.. Solution_2 o..
    # Pythonic way.  recursively.
    # According to: https://leetcode.com/discuss/42550/one-liners-in-python
    ___ permute  nums
        r_ [[n] + p
                ___ i, n __ enumerate(nums)
                ___ p __ self.permute(nums[:i] + nums[i + 1:])] or [[]]

"""
[]
[1]
[1,2,3]
"""
